x/sync/README.md

```diff
@@ -1,3 +1,57 @@
+
+This message is sent from the client to the server to request a range proof for a given key range and root ID. That is, the client says, "Give me the key-value pairs that were in this key range when the database had this root."
+
+### `RangeProof`
+
+This message is sent from the server to the client in response to a `SyncGetRangeProofRequest`. It contains the key-value pairs that were in the requested key range when the database had the requested root, as well as a proof that the key-value pairs are correct.
+
+### `SyncGetChangeProofRequest`
+
+This message is sent from the client to the server to request a change proof between the given root IDs. That is, the client says, "Give me the key-value pairs that changed between the time the database had this root and that root."
+
+### `SyncGetChangeProofResponse`
+
+This message is sent from the server to the client in response to a `SyncGetChangeProofRequest`. If the server had sufficient history to generate a change proof, it contains a change proof that contains the key-value pairs that changed between the requested roots. If the server did not have sufficient history to generate a change proof, it contains a range proof that contains the key-value pairs that were in the database when the database had the latter root.
+
+## Algorithm
+
+When the client starts with an empty database, it has no key-value pairs. It requests from a server a range proof for the entire database. The server replies with a range proof, which the client verifies. If it's valid, the key-value pairs in the proof are written to the database. If it's not, the client drops the proof and requests the proof from another server. 
+
+A range proof sent by a server may be valid but not contain all of the key-value pairs in the requested range. For example, a client might request all the key-value pairs in [`start`, `end`] but only receive those in range [`start`, `end'`] where `end'` < `end`. There might be too many key-value pairs to include in one message, or the server may be too busy to provide any more in its response. Unless the database is very small, this means that the range proof the client receives in response to its range proof request for the entire database will not contain all of the key-value pairs in the database.
+
+For each key range, the sync client keeps track of the root ID of the last time it synced that key range. For example, it will store information that says something like, "I have all of the key-value pairs that were in range [`start`, `end`] when the database's root was `rootID`" for some keys `start` and `end`, and some database `rootID`.
+
+If the first range proof contained all of the key-value pairs up to some key `end`, the client recognizes that it must fetch all of the keys after `end`. It divides the remaining key range into chunks, and repeats this process by asking servers for a range proof for each chunk until it has all of the key-value pairs in the database.
```

Comment:
> It divides the remaining key range into chunks,

not sure it is worth explicitly mentioning, but it doesn't always divide and it also prioritizes the first chunk over the second chunk so that ranges with matching rootID will grow rather than have many disjoint ranges.

Reply:
reworded it a bit. I think the exact algorithm for chunking/dechunking (those are words, right?) is too much detail for this document since a client could choose to implement it differently